File: 8.Sorting/GoLang/4_merge_sort.go
package main

func show(arr []int, msg string) {
	println(msg)
	for i := 0; i < len(arr); i++ {
		print(arr[i], " ")
	}
	println()
}

func merge(arr []int, start int, mid int, end int) {
	l := mid - start + 1
	r := end - mid
	templ := make([]int, l)
	tempr := make([]int, r)

	for i := 0; i < l; i++ {
		templ[i] = arr[start+i]
	}
	for i := 0; i < r; i++ {
		tempr[i] = arr[mid+i+1]
	}
	i := 0
	j := 0
	k := start
	for i < l && j < r {
		if templ[i] <= tempr[j] {
			arr[k] = templ[i]
			i++
		} else {
			arr[k] = tempr[j]
			j++
		}
		k++
	}

	for i < l {
		arr[k] = templ[i]
		i++
		k++
	}
	for j < r {
		arr[k] = tempr[j]
		i++
		j++
	}

}

func mergesort(arr []int, start int, end int) {
	if start < end {
		var mid int = start + (end-start)/2
		mergesort(arr, start, mid)
		mergesort(arr, mid+1, end)
		merge(arr, start, mid, end)
	}

}

func main() {
	arr := []int{22, 9, 3, 55, 43, 0, 1}
	show(arr, "before")
	mergesort(arr, 0, len(arr)-1)
	show(arr, "after")

}

File: 8.Sorting/Java/4_merge_sort.java
class MergeSort {
    void show(int arr[], String msg) {
        System.out.println(msg);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    void merge(int arr[], int start, int mid, int end) {
        int l = mid - start + 1;
        int r = end - mid;
        int templ[] = new int[l];
        int tempr[] = new int[r];
        for (int i = 0; i < templ.length; i++) {
            templ[i] = arr[start + i];
        }
        for (int i = 0; i < tempr.length; i++) {
            tempr[i] = arr[mid + i + 1];
        }

        int i = 0;
        int j = 0;
        int k = start;

        while (i < templ.length && j < tempr.length) {
            if (templ[i] < tempr[j]) {
                arr[k] = templ[i];
                i++;
            } else {
                arr[k] = tempr[j];
                j++;
            }
            k++;
        }

        while (i < templ.length) {
            arr[k] = templ[i];
            i++;
            k++;
        }

        while (j < tempr.length) {
            arr[k] = tempr[j];
            j++;
            k++;
        }
    }

    void mergeSort(int arr[], int start, int end) {
        if (start < end) {
            int mid = start + (end - start) / 2;
            mergeSort(arr, start, mid);
            mergeSort(arr, mid + 1, end);
            merge(arr, start, mid, end);
        }
    }

    public static void main(String arg[]) {
        int arr[] = {22, 1, 9, 46, 0, 55, 3};
        MergeSort o = new MergeSort();
        o.show(arr, "before");
        o.mergeSort(arr, 0, arr.length - 1);
        o.show(arr, "after");

    }
}
